### DISCUSSION TITLE
Suggestions for lazer editor

### DISCUSSION THREAD

**Discussion by iceCreamMint** [ppy/osu#21229] [Ideas]
The following is a list of possible editor improvements that i think would make a better lazer editor, subjectively sorted in significance. There are also some bugs/glitches at the end
2, 3, 4, 5, 8 and the need for a larger editor area can probably solved with a skin editor for the editor, but these changes will probably make for a better default editor.
I dont think these are duplicates of anything, but if any of them are, I apologize.

1. anchor types are confusing; one of the biggest examples of this is that when you double click it creates a linear anchor, but then you add an inherited node afterwards and the previous node becomes a perfect curve, kind of unintuitive. a very common thing that happens is that i want to change a 3 point slider into a bezier, and my intuition tells me to change the type of the middle curve(because intuitively it "defines" the curve), which effectively turns it into two linear segments

2. have object ghosts be same size as circles in skin, cuz it makes spacing wonky when objects in the future are visually different sizes from objects in the past

3. timelines are completely opaque so stuff is cut off from top and bottom of screen. also being able to see the entirety of your bg when mapping is nice

4. bottom timeline takes up too much space, there are less things than stable and the stuff that remain are more compact, yet it takes up more space. this and 3 will also resolve the need for a larger editing area so that wont be its separate entry in the list

5. object fade out timing is quicker than stable, and the object ghosts for sliderheads and ends have way too much presence compared to sliderbodies, though this may be due to how quickly sliderbodies fade out

6. have a baseline sv control to avoid situations where the entire map is >2.0 sv, unless im missing something theres no way to set it

7. its possible to create a slider entirely on screen even though the anchors are offscreen, and sometimes this is neccesary for large bezier sliders, but currently in lazer you have to finish placing the slider and then move the anchor outside, which is annoying

8. change the object ghosts to be some effect/filter on the object instead of a filled white circle, the mapping skin i use has transparent circles and the basically opaque circles throw me off

9. remove object, which is a really common action, now takes either two clicks or moving your hand across the keyboard to the delete key, so having it on the left of the keyboard in the area where we switch between objects would be nice

10. add more hotkey settings, or at least a help menu to show what hotkeys there are, im not sure if theres a slider to stream hotkey but if there is i cant find it. would also solve 9

11. the ability to remove, add, and change the length of breaks is used in some maps. currently theres no way to change breaks, if you place objects on a break they exist through the break, and if you place two objects far apart a break is not created(i guess those last two are bugs)

12. bookmarks would be nice

13. no way to set sv by section makes sense, because you need to adjust each slider anyways, but i think changing volumes by section makes sense, so add back green lines or something similar

14. add support for different hitnormals and additions, as well as having different samples on sliderhead, sliderend, each individual repeat, etc. and because of custom samples the ability to access song folder from editor will be neccessary(at the very least i cant find how to do this)

15. in stable when you place an object on top of another object it deletes the previous object(usually) which is useful. in lazer it doesnt which is great for 2b but not really for normal maps

list of bugs/glitches i got:
editor lags when saving maps, quite inconvenient when i like to spam save every few objects
the sliderball becomes normal to slider body when moving an anchor, looks weird on skins with arrows like argon
changing volume for a slider changes its sampleset to normal, this also happens when changing sampleset, and this only happens with sliders, circles are fine
changing anchor type does not recalculate slider length
cannot ctrl+z sv and volume changes

**Comment by GopoBoi** [comment]
> have object ghosts be same size as circles in skin, cuz it makes spacing wonky when objects in the future are visually different sizes from objects in the past

The white part is the actual size, you can check this by using one of the new default skins.

> bottom timeline takes up too much space, there are less things than stable and the stuff that remain are more compact, yet it takes up more space. this and 3 will also resolve the need for a larger editing area so that wont be its separate entry in the list

You can adjust the UI scaling to make the elements smaller. this will make the playfield take up more space.

> have a baseline sv control to avoid situations where the entire map is >2.0 sv, unless im missing something theres no way to set it

Sliders inherit the previous one's sv.

> remove object, which is a really common action, now takes either two clicks or moving your hand across the keyboard to the delete key, so having it on the left of the keyboard in the area where we switch between objects would be nice

Shift + right click removes an object.

> no way to set sv by section makes sense, because you need to adjust each slider anyways, but i think changing volumes by section makes sense, so add back green lines or something similar

you can set any of the attributes to whole selections.

**Reply by iceCreamMint** [reply]
> The white part is the actual size, you can check this by using one of the new default skins.

yeah but visually even the size of sliderbodies in argon agree with the smaller circle sizes
<img width="380" alt="image" src="https://user-images.githubusercontent.com/59371178/201675862-45781459-24f5-4d61-91e8-31f38dc74708.png">
in mapping its more important that its visually consistent rather in a skin rather than to true circle size(hot take) so this is either solved by changing skinning conventions or some variation to 8, where the size of the object ghost is dependent on skin. could even add an indicator to reference true size, like argon where the true circle size is a ring around the object

> You can adjust the UI scaling to make the elements smaller. this will make the playfield take up more space.

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/59371178/201677566-42f9ffaf-8614-4e8f-95a8-b85c60028bf6.png">
this is the smallest i can set it to from settings(0.8), and the bottom timeline is still very plain and contains basically no information, even the "red line" indicators are smaller than in stable. and at the smallest ui scaling i can put it to, it is still larger than stable on my screen

> Sliders inherit the previous one's sv.

but the max slider velocity is still 10x, which means effective max slider velocity is lower than in stable, case in point this 1.0x sv slider in stable
<img width="1280" alt="image" src="https://user-images.githubusercontent.com/59371178/201679357-fce47907-67c9-4022-8d12-2f35cfcc4db5.png">
as compared to lazer
<img width="1280" alt="image" src="https://user-images.githubusercontent.com/59371178/201679475-fcb3224d-343a-43d5-9d80-3b6fbf7256ba.png">

> Shift + right click removes an object.

thanks, but also indicates the need for a view hotkey menu as pointed out in 10

> you can set any of the attributes to whole selections.

fair enough, i guess, but right now its difficult to view the attributes of an entire section at a glance. if you want to change the volume for an entire section, you need to look for the start and end of that section in the top timeline, which becomes inconvinient
